File: app/src/main/java/com/bignerdranch/stockwatcher/model/service/StockService.java
package com.bignerdranch.stockwatcher.model.service;


import java.util.List;

import io.reactivex.Observable;
import retrofit2.Retrofit;
import retrofit2.http.GET;
import retrofit2.http.Query;

public class StockService {

    private final StockServiceInterface service;

    interface StockServiceInterface {
        @GET("Quote/json")
        Observable<StockInfoResponse> stockInfo(@Query("symbol") String symbol);

        @GET("Lookup/json")
        Observable<List<StockSymbol>> lookupStock(@Query("input") String symbol);
    }

    public StockService(Retrofit retrofit) {
        service = retrofit.create(StockServiceInterface.class);
    }

    public Observable<StockInfoResponse> stockInfo(String symbol) {
        return service.stockInfo(symbol);
    }

    public Observable<StockSymbol> lookupStock(String symbol) {
        return service.lookupStock(symbol)
                .concatMap(Observable::fromIterable)
                .firstOrError()
                .toObservable();
    }
}

File: app/src/main/java/com/bignerdranch/stockwatcher/model/service/StockService.kt
package com.bignerdranch.stockwatcher.model.service


import io.reactivex.Observable
import retrofit2.Retrofit
import retrofit2.http.GET
import retrofit2.http.Query

class StockService(retrofit: Retrofit) {

    private val service: StockServiceInterface

    init {
        service = retrofit.create(StockServiceInterface::class.java)
    }

    internal interface StockServiceInterface {
        @GET("Quote/json")
        fun stockInfo(@Query("symbol") symbol: String): Observable<StockInfoResponse>

        @GET("Lookup/json")
        fun lookupStock(@Query("input") symbol: String): Observable<List<StockSymbol>>
    }

    fun stockInfo(symbol: String): Observable<StockInfoResponse> {
        return service.stockInfo(symbol)
    }

    fun lookupStock(symbol: String): Observable<List<StockSymbol>> {
        return service.lookupStock(symbol)
    }
}
